cou/apps/app.py

```diff
@@ -245,24 +224,121 @@ def __str__(self) -> str:
             yaml.dump(summary, stream)
             return stream.getvalue()
 
+    def _populate_units(self) -> None:
+        """Populate application units."""
+        if not self.is_subordinate:
+            for name, unit in self.status.units.items():
+                compatible_os_version = (
+                    self._get_latest_compatible_os_version_by_unit_workload_version(unit)
+                )
+                self.units.append(
+                    ApplicationUnit(
+                        name=name,
+                        workload_version=unit.workload_version,
+                        os_version=compatible_os_version,
+                        machine=unit.machine,
+                    )
+                )
+
+    def _get_latest_compatible_os_version_by_unit_workload_version(
+        self, unit: UnitStatus
+    ) -> OpenStackRelease:
+        """Get the latest compatible OpenStack release based on the unit workload version.
+
+        If the application is versionless, then get from the current OpenStack release.
+
+        :param unit: Application Unit
+        :type unit: UnitStatus
+        :raises ApplicationError: When there are no compatible OpenStack release for the
+        workload version.
+        :return: The latest compatible OpenStack release.
+        :rtype: OpenStackRelease
+        """
+        if self.is_os_channel_based:
+            return self.channel_codename
+
+        try:
+            return max(
+                OpenStackCodenameLookup.find_compatible_versions(self.charm, unit.workload_version)
+            )
+        except ValueError as exc:
+            raise ApplicationError(
+                f"'{self.name}' with workload version {unit.workload_version} has no "
+                f"compatible OpenStack release."
+            ) from exc
+
+    @property
+    def channel(self) -> str:
+        """Get charm channel of the application.
+
+        :return: Charm channel. E.g: ussuri/stable
+        :rtype: str
+        """
+        return self._channel
+
+    @channel.setter
+    def channel(self, charm_channel: str) -> None:
+        """Set charm channel of the application.
+
+        When application comes from charm store, the channel won't be OpenStack related.
+        :param charm_channel: Charm channel. E.g: ussuri/stable
+        :type charm_channel: str
+        :raises ValueError:  Exception raised when channel is not a valid OpenStack
+            channel.
+        """
+        if (
+            self.is_from_charm_store
+            or (not self.is_os_channel_based and self.is_subordinate)
+            or self.is_valid_track
+        ):
+            self._channel = charm_channel
+            return
+        raise ValueError(self.channel_err_msg(charm_channel))
```

Comment:
not-blocking: 
```suggestion
        if not (
            self.is_from_charm_store
            or (not self.is_os_channel_based and self.is_subordinate)
            or self.is_valid_track
        ):
            raise ValueError(self.channel_err_msg(charm_channel))

        self._channel = charm_channel
```

Reply:
IMO using `not`make it less readable

Reply:
To me it's always more readable to do break, raise or return None/default value at the beginning of function. To separate actual logic of function. The `self._channel = charm_channel` is main purpose of this function and raising exception is corner case.
But it's only my point of view and it's not blocking. Resolving this comment.